# Проект. Ещё больше о Pandas.
#  «Аналитика бронирования отелей»

## Шаг 1

В этом проекте вам предстоит почувствовать себя в роли аналитика данных в компании, владеющей сетью отелей, и проанализировать данные о бронированиях. Описание колонок:
* Hotel – тип отеля (City Hotel или Resort Hotel)  
* Is canceled – бронирование было отменено (1) или нет (0); не отмененное считается успешным
* Lead time – количество дней, прошедших между датой бронирования и датой прибытия  
* Arrival full date – полная дата прибытия
* Arrival date year – год прибытия  
* Arrival date month – месяц прибытия  
* Arrival date week number – номер недели прибытия
* Arrival date day of month – день прибытия
* Stays in weekend nights – количество выходных (суббота или воскресенье), которые гость забронировал для проживания в отеле
* Stays in week nights – количество дней (с понедельника по пятницу), которые гость забронировал для проживания в отеле
* Stays total nights – общее число забронированных ночей (сумма двух предыдущих колонок)
* Adults – число взрослых
* Children – число детей
* Babies – число младенцев
* Meal – выбранный тип питания
* Country – страна происхождения клиента
* Reserved room type – тип зарезервированного номера
* Assigned room type – тип полученного номера (может отличаться от забронированного)
* Customer type – тип бронирования
* Reservation status – значение последнего статуса брони: Canceled — было отменено клиентом; Check-Out — клиент зарегистрировался, но уже покинул отель; No-Show — клиент не зарегистрировался и сообщил администрации отеля причину
* Reservation status date – дата обновления статуса

В этом уроке вы закрепите навыки решения аналитических задач с помощью основных методов библиотеки pandas и попробуете новые методы.

Вы можете решать задания тем способом, который считаете наиболее удобным или оптимальным. А для того, чтобы лучше понять задание, следуйте по шагам выполнения. Бывает, что их можно выполнить в одну строчку, применяя методы друг за другом. А если выполняете шаги отдельно, не забудьте сохранить результат в переменную, чтоьбы в следующем шаге работать именно с ней.

В ноутбуке место для вашего решения обозначено комментарием `# Ваш код здесь`, но вы можете писать код там, где вам удобно, добавлять или удалять ячейки с кодом или текстом по вашему усмотрению.

Напоминаем и о [памятке для новичков в программировании](https://buildin.ai/share/62b112ad-e20b-4b85-ab63-877077d1e9c3?code=2J58AF
)! Очень многие проблемы, описанные в ней, часто встречаются именно в этом уроке, поэтому рекомендуем открыть её в отдельном окне и периодически с ней сверяться.  

Кроме того можно посмотреть подсказки, кликнув на строчку `► Нажмите сюда, чтобы увидеть подсказку`. Также не забывайте о наличии конспектов и возможности задать вопрос в Пачку, ссылка на нужный тред есть на странице каждого шага в LMS.

## Шаг 2

Как обычно, в начале откройте датасет и посмотрите на данные. Для этого используйте библиотеку pandas, прочитайте датасет с разделителем `;`, сохраните его в переменную `bookings`. **Посмотрите, из какой страны было сделано бронирование в 19-й сверху строке** (обратите внимание, индекс у этой строки может быть не 19)

Шаги выполнения:

1. Импортируйте библиотеку pandas с общепринятым алиасом (псевдонимом) pd

In [1]:
import pandas as pd

2. Загрузите csv файл к уроку в свою папку в JupyterНub

3. Прочитайте этот датасет с разделителем `;`, результат сохранив в переменную `bookings`

In [2]:
bookings = pd.read_csv('bookings.csv', sep=';')

4. Выведите первые 19 строк датафрейма

In [3]:
bookings.head(19)['Country']

0     PRT
1     PRT
2     GBR
3     GBR
4     GBR
5     GBR
6     PRT
7     PRT
8     PRT
9     PRT
10    PRT
11    PRT
12    USA
13    ESP
14    PRT
15    IRL
16    PRT
17    IRL
18    FRA
Name: Country, dtype: object

5. Определите название страны из 19-й сверху строки в колонке `Country`

**Ответ: FRA**

## Шаг 3

Посмотрите на размеры датафрейма `bookings`. **Сколько всего в нем строк и столбцов?**

In [4]:
bookings.shape

(119390, 21)

**Ответ:  
число строк: 119390  
число столбцов: 21**

## Шаг 4

**К какому типу/каким типам относится большинство столбцов?** Выберите один или несколько вариантов.

Шаги выполнения:
1. Проверьте типы данных в датафрейме

In [5]:
bookings.dtypes

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

2. Определите, какой тип/какие типы данных преобладают

In [6]:
bookings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 21 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Hotel                      119390 non-null  object 
 1   Is Canceled                119390 non-null  int64  
 2   Lead Time                  119390 non-null  int64  
 3   arrival full date          119390 non-null  object 
 4   Arrival Date Year          119390 non-null  int64  
 5   Arrival Date Month         119390 non-null  object 
 6   Arrival Date Week Number   119390 non-null  int64  
 7   Arrival Date Day of Month  119390 non-null  int64  
 8   Stays in Weekend nights    119390 non-null  int64  
 9   Stays in week nights       119390 non-null  int64  
 10  stays total nights         119390 non-null  int64  
 11  Adults                     119390 non-null  int64  
 12  Children                   119386 non-null  float64
 13  Babies                     11

**Ответ: int64, object**

## Шаг 5

Как вы уже могли заметить, названия колонок записаны в неудобном для дальнейшей работы формате. **Исправьте ситуацию, приведя названия столбцов к нижнему регистру и заменив пробелы на знак нижнего подчеркивания**.
Например: `Is Canceled` -> `is_canceled`

Вручную это делать долго и неудобно, попробуйте автоматизировать ваше решение. Для этого могут пригодиться:

[rename](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html)  
[str.lower](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.lower.html)  
[str.replace](https://docs.python.org/3/library/stdtypes.html#str.replace)  
[columns](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.columns.html)

Подробнее о них можно прочитать в документации по ссылкам или в конспектах этого и предыдущего уроков.

Шаги выполнения:
1. На одном названии колонки потренируйтесь приводить его к нижнему регистру и заменять пробелы на знак нижнего подчеркивания. Для примера можете взять название колонки `Is Canceled` как строку, в кавычках.

In [7]:
'Is Canceled'.replace(' ', '_').lower()

'is_canceled'

2. Теперь используйте получившийся код для всех названий колонок. Пройдя циклом по списку колонок датафрейма, соберите словарь, в котором ключами будут старые названия колонок, а значениями — новые. Перед циклом, но желательно в той же ячейке, не забудьте создать пустой словарь, куда будете складывать пары названий колонок ключ-значение.

In [8]:
rename_cols = {}
for col in bookings.columns:
    rename_cols[col] = col.replace(' ', '_').lower()

3. Убедитесь, что в словаре с названиями колонок получились именно те значения, которые вам нужны

In [9]:
rename_cols

{'Hotel': 'hotel',
 'Is Canceled': 'is_canceled',
 'Lead Time': 'lead_time',
 'arrival full date': 'arrival_full_date',
 'Arrival Date Year': 'arrival_date_year',
 'Arrival Date Month': 'arrival_date_month',
 'Arrival Date Week Number': 'arrival_date_week_number',
 'Arrival Date Day of Month': 'arrival_date_day_of_month',
 'Stays in Weekend nights': 'stays_in_weekend_nights',
 'Stays in week nights': 'stays_in_week_nights',
 'stays total nights': 'stays_total_nights',
 'Adults': 'adults',
 'Children': 'children',
 'Babies': 'babies',
 'Meal': 'meal',
 'Country': 'country',
 'Reserved Room Type': 'reserved_room_type',
 'Assigned room type': 'assigned_room_type',
 'customer type': 'customer_type',
 'Reservation Status': 'reservation_status',
 'Reservation status_date': 'reservation_status_date'}

4. Используя получившийся словарь, переименуйте колонки датафрейма `bookings`, сохранив результат в том же датафрейме

In [10]:
bookings = bookings.rename(columns=rename_cols)

5. Убедитесь, что названия столбцов были успешно изменены и результат сохранен в датафрейм `bookings`

In [11]:
bookings.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_full_date',
       'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'stays_total_nights', 'adults', 'children',
       'babies', 'meal', 'country', 'reserved_room_type', 'assigned_room_type',
       'customer_type', 'reservation_status', 'reservation_status_date'],
      dtype='object')

## Шаг 6

Перейдем к исследованию данных. Чтобы вести успешный бизнес, нашей сети отелей нужно понимать своих клиентов. В этом задании выясните, пользователи из каких стран совершили наибольшее число успешных бронирований? Бронирование считается успешным, если в дальнейшем не было отменено (см.колонку `is_canceled`). **В качестве ответа выберите страны, входящие в топ-5 по числу успешных бронирований**.

Шаги выполнения:  
1. Выберите только те записи, где бронирование не было отменено (значение в колонке `is_canceled` равно 0)

In [12]:
bookings.query('is_canceled==0').country.value_counts().head()

country
PRT    21071
GBR     9676
FRA     8481
ESP     6391
DEU     6069
Name: count, dtype: int64

2. Посчитайте количество оставшихся после фильтрации бронирований для каждой страны

3. На основе получившихся расчетов определите 5 стран с наибольшим числом успешных бронирований

**Ответ: DEU, ESP, FRA, GBR, PRT**

## Шаг 7

В нашей сети есть два типа отелей: City Hotel и Resort Hotel. Владельцы сети хотят понимать, отличается ли длительность бронирования в зависимости от типа отеля (см.колонку `hotel`). **Выясните, на сколько ночей (см.колонку `stays_total_nights`) в среднем бронируют отели City Hotel и на сколько Resort Hotel?** Запишите полученные значения в пропуски с точностью до двух знаков после точки.

Может пригодиться: [round](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.round.html)

Шаги выполнения:  
1. Сгруппируйте данные по типу отеля и посчитайте для каждой группы среднее число забронированных ночей

In [13]:
bookings.groupby('hotel').stays_total_nights.mean().round(2)

hotel
City Hotel      2.98
Resort Hotel    4.32
Name: stays_total_nights, dtype: float64

2. Округлите результаты до двух знаков после точки

**Ответ:  
Среднее число забронированных ночей в отелях типа City Hotel: 2.98  
В Resort Hotel: 4.32**

## Шаг 8

Бывает, что в наших отелях случается [овербукинг](https://ru.wikipedia.org/wiki/Овербукинг). Тогда тип номера, присвоенного клиенту (см.колонку `assigned_room_type`), будет отличаться от изначально забронированного (см.колонку `reserved_room_type`). **Посчитайте, сколько подобных наблюдений встретилось в датасете?** Отмена бронирования также считается.

Шаги выполнения:
1. Отберите строки, где значения в колонке `assigned_room_type` не совпадают со значениями в колонке `reserved_room_type`

In [14]:
bookings.query('assigned_room_type != reserved_room_type').shape

(14917, 21)

2. Посчитайте количество оставшихся строк

**Ответ: 14917**

## Шаг 9

Если у спроса на услугу есть сезонность, компании стоит это учесть. Проанализируйте даты запланированного прибытия (см.колонку `arrival_date_year`) и **определите, на какой месяц чаще всего оформляли бронь в 2016 году? Изменился ли самый популярный месяц в 2017?**

Шаги выполнения:
1. Сгруппируйте данные по годам и посчитайте, сколько раз внутри года встречается каждый месяц

In [15]:
bookings.groupby('arrival_date_year').arrival_date_month.value_counts()

arrival_date_year  arrival_date_month
2015               September             5114
                   October               4957
                   August                3889
                   December              2920
                   July                  2776
                   November              2340
2016               October               6203
                   May                   5478
                   April                 5428
                   September             5394
                   June                  5292
                   August                5063
                   March                 4824
                   July                  4572
                   November              4454
                   February              3891
                   December              3860
                   January               2248
2017               May                   6313
                   April                 5661
                   June                  5

2. Посмотрите, в каком месяце было больше всего бронирований в 2016 году. А в 2017?

**Ответ:  
В 2016 чаще всего бронировали отель на октябрь.  
В 2017 чаще всего оформляли бронирование на май**.

## Шаг 10

Продолжим изучать бронирования. **Выясните, в каком месяце каждого года бронирования отеля типа City Hotel отменялись чаще всего.**

Шаги выполнения:
1. Оставьте только записи об отмененных бронированиях (см.колонку `is_canceled`) отеля типа City Hotel (см.колонку `hotel`)

In [16]:
bookings.query('hotel == "City Hotel" and is_canceled == 1') \
    .groupby('arrival_date_year') \
    .arrival_date_month \
    .value_counts()

arrival_date_year  arrival_date_month
2015               September             1543
                   October               1321
                   August                1232
                   July                   939
                   December               668
                   November               301
2016               October               1947
                   June                  1720
                   September             1567
                   April                 1539
                   May                   1436
                   November              1360
                   August                1247
                   March                 1108
                   December              1072
                   July                  1043
                   February               930
                   January                438
2017               May                   2217
                   April                 1926
                   June                  1

2. Сгруппируйте оставшиеся данные по годам и посчитайте количество записей для каждого месяца внутри года

3. Посмотрите, в каком месяце бронирования City Hotel отменялись чаще всего в 2015? 2016? 2017?

**Ответ:  
2015 - September  
2016 - October  
2017 - May**

## Шаг 11

Теперь посмотрим на наших клиентов с точки зрения их возраста. **Какая из колонок `adults`, `children` и `babies` имеет наибольшее среднее значение?**

Шаги выполнения:
1. Оставьте в датафрейме только 3 колонки:  `adults`, `children` и `babies`

In [17]:
bookings[['adults', 'children', 'babies']].mean()

adults      1.856403
children    0.103890
babies      0.007949
dtype: float64

2. Посчитайте средние значения для оставшихся колонок

3. Посмотрите, среднее значение в какой колонке больше других?

**Ответ: adults**

## Шаг 12

В наших данных информация о количестве детей записана в две колонки: `children` и `babies`. Это удобно, если нам нужно знать примерный возраст маленьких гостей, но неудобно, если нам требуется просто общее число детей. Для этого создайте колонку `total_kids`, объединив столбцы `children` и `babies`.

После того проверьте, для отелей какого типа среднее значение этой переменной оказалось наибольшим, для City hotel или Resort hotel. **В качестве ответа укажите наибольшее среднее `total_kids`, округлив до двух знаков после точки**.

Шаги выполнения:
1. Создайте колонку `total_kids`, сложив значения в столбцах `children` и `babies`

In [18]:
bookings['total_kids'] =  bookings.children + bookings.babies

2. Сгруппируйте данные по типу отеля и посчитайте для каждой группы среднее количество детей

In [19]:
bookings.groupby('hotel').agg({'total_kids':'mean'}).round(2)

total_kids
hotel                   
City Hotel          0.10
Resort Hotel        0.14

3. Округлите результаты до двух знаков после точки

4. Посмотрите, для отелей какого типа среднее значение оказалось наибольшим? Чему равно это значение?

**Ответ: 0.14**

## Шаг 13

Бизнес не любит упускать прибыль, но не все бронирования завершились успешно (см.колонку `is_canceled`), поэтому попробуем посчитать, сколько клиентов было потеряно в процессе. Иными словами, посчитаем метрику под названием Churn Rate.

**Churn rate** (отток, коэффициент оттока) – это процент подписчиков, которые отписались от канала коммуникации, отказались от услуг сервиса в течение определенного периода времени. Т.е. представляет собой отношение количества ушедших пользователей к общему количеству пользователей, выраженное в процентах. Если Churn rate считают для какой-то определенной группы, то и количество ушедших пользователей, и общее количество пользователей считают только по этой группе.

В нашем случае Churn Rate — это процент клиентов, которые отменили бронирование. Давайте посмотрим, как эта метрика связана с наличием детей у клиентов. **Посчитайте Churn Rate дважды: для клиентов с детьми и для клиентов без детей. В качестве ответа укажите наибольший Churn Rate (% оттока), округленный до двух знаков после точки** (то есть доля 0.24563 будет 24.56% и в ответ пойдёт 24.56)

Шаги выполнения:  
1. Создайте колонку `has_kids`, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (см.созданную на прошлом шаге колонку `total_kids`), в противном случае – False

In [20]:
bookings['has_kids'] = bookings['total_kids'] >= 1

2. Получите 4 датафрейма, отобрав только нужные записи для каждого:  
1 — клиенты без детей, которые делали бронирование (не важно, отменяли его потом или нет)  
2 — клиенты без детей, которые отменили бронирование  
3, 4 — аналогичные два датафрейма для клиентов с детьми

3. Посчитайте количество записей в каждом из четырех датафреймов, чтобы получить число клиентов с теми или иными параметрами

4. Посчитайте Churn Rate клиентов без детей: разделите число клиентов без детей, которые отменили бронирование, на общее число клиентов без детей и переведите результат в проценты. По аналогии посчитайте Churn Rate клиентов с детьми

In [21]:
# Churn Rate клиентов без детей
round(bookings.query('has_kids==False and is_canceled==1').shape[0] / bookings.query('has_kids==False').shape[0] * 100, 2)

37.22

5. Округлите оба значения Churn Rate до двух знаков после точки

6. Посмотрите, какое из значений Churn rate больше: у клиентов с детьми или у клиентов без детей? Чему равно это значение?

In [22]:
# Churn Rate клиентов с детьми
round(bookings.query('has_kids==True and is_canceled==1').shape[0] / bookings.query('has_kids==True').shape[0] * 100, 2)

34.92

**Ответ: 37.22**

## Итоги

В этом проекте вы проанализировали данные о бронированиях и закрепили навыки работы с библиотекой pandas, а также узнали новые подходы к переименованию колонок и округлению значений, познакомились с метрикой Churn Rate.